explainer.md

```diff
@@ -22,118 +71,260 @@ place associated with the application (such as the shelf or home screen).
 
-TODO: An issue is that if the methods are called from a service worker whose
-scope is a parent of the web app manifest scope, it would be ambiguous which web
-app is being identified. We need to take an optional scope parameter.
+### Examples
 
-Example code (from the main window):
+#### Basic Examples
+These examples provided limited context. They are intended to show how the API will look.
 
-Setting an integer badge (as in an email app):
+##### Setting an integer badge (as in an email app):
 ```js
 Badge.set(getUnreadCount());
 ```
 
-Setting and clearing a boolean flag (as in a game of chess):
+##### Setting and clearing a boolean flag (as in a game of chess):
 ```js
 if (myTurn())
   Badge.set();
 else
   Badge.clear();
 ```
 
+#### Advanced Examples
+These examples generally provide more context around what is going on, and may specify scopes. They are intended to show how real world applications may use the API.
+
+##### Updating a Badge on a message from a WebSocket (as in a messaging app, receiving new messages):
+
+```ts
+interface ChatMessage {
+  content: string;
+  unread: boolean;
+}
+
+// A list of all messages received by the application.
+const messages: ChatMessage[] = [];
+
+// Sets the badge count to the number of unread messages.
+const updateBadgeCount = () => {
+  // Count the number of unread messages.
+  const badgeCount = messages
+    .filter(m => m.unread)
+    .length;
+  Badge.set(badgeCount);
+}
+
+// Adds a new message and updates the badge count.
+const addMessage = (message: Message) => {
+  messages.push(message);
+  updateBadgeCount();
+}
+
+// Create a new socket, connected to our messages endpoint.
+const socket = new WebSocket('ws://example.com/messages');
+socket.onmessage = (event) => {
+  // Parse the message and add it to our list.
+  const chatMessage = JSON.parse(event.data);
+  addMessage(chatMessage);
+};
+```
+
+##### Setting a separate badge for the app and a specific page (as in the case of github notifications an PR statuses).
+On all pages of this site, we wish to display the notification count, except for `/do-have-https`, where we should instead display `flag` if the page was not served over `https` or nothing, if it was.
+
+The main page of our site https://example.com/
+```html
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <title>Main Page</title>
+  </head>
+  <body>
+    <script>
+      // There are ten unread notifications.
+      // This was injected by the server.
+      Badge.set(10);
+    </script>
+  </body>
+</html>
+```
+
+On https://example.com/do-i-have-https
+```html
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <title>Main Page</title>
+  </head>
+  <body>
+    <script>
+      const badgeOptions = {
+        scope: `/do-i-have-https`
+      };
+
+      const hasHttps = window
+        .location
+        .href
+        .startsWith('https');
+
+      // Note: This is not supported.
+      Badge.set(hasHttps ? '✓' : '☠', badgeOptions);
```

Comment:
This doesn't conform with the IDL yet... which only allows `unsigned long long`.

Reply:
See the `Note:` in the comment. Do you think I should completely remove it?

Reply:
Lol, sorry... totally missed the comment! :P yeah, I think we should remove it for now.

Reply:
Cool, done :)